pints/_transformation.py

```diff
@@ -0,0 +1,574 @@
+# released under the BSD 3-clause license. See accompanying LICENSE.md for
+# copyright notice and full license details.
+#
+from __future__ import absolute_import, division
+from __future__ import print_function, unicode_literals
+import pints
+import numpy as np
+from scipy.special import logit, expit
+
+
+class Transform(object):
+    """
+    Abstract base class for objects that provide some convenience parameter
+    transformation from the model parameter space to a search space.
+
+    If ``t`` is an instance of a ``Transform`` class, you can apply
+    the transformation of a parameter vector from the model space ``p`` to the
+    search space ``x`` by using ``x = t.to_search(p)`` and the inverse by using
+    ``p = t.to_model(x)``.
+    """
+    def apply_log_pdf(self, log_pdf):
+        """
+        Returns a transformed log-PDF class.
+        """
+        return TransformedLogPDF(log_pdf, self)
+
+    def apply_error_measure(self, error_measure):
+        """
+        Returns a transformed error measure class.
+        """
+        return TransformedErrorMeasure(error_measure, self)
+
+    def apply_boundaries(self, boundaries):
+        """
+        Returns a transformed boundaries class.
+        """
+        return TransformedBoundaries(boundaries, self)
+
+    def jacobian(self, x):
+        """
+        Returns the Jacobian for a parameter vector ``x`` in the search space.
+
+        *This is an optional method. It is needed when `sigma0` is provided in
+        :meth:`pints.OptimisationController` or :meth:`pints.MCMCController`,
+        or when the method requires ``evaluateS1()``.*
+        """
+        raise NotImplementedError
+
+    def log_jacobian_det(self, x):
+        """
+        Returns the logarithm of the absolute value of the Jacobian
+        determinant for a parameter vector ``x`` in the search space.
+
+        *This is an optional method. It is needed when transformation is
+        performed on :class:`LogPDF` and/or that requires ``evaluateS1()``;
+        e.g. not necessary if it's used for :class:`ErrorMeasure` without
+        :meth:`ErrorMeasure.evaluateS1()`.*
+        """
+        return np.log(np.abs(np.linalg.det(self.jacobian(x))))
+
+    def n_parameters(self):
+        """
+        Returns the dimension of the parameter space this transformation is
+        defined over.
+        """
+        raise NotImplementedError
+
+    def to_model(self, x):
+        """
+        Returns the inverse of transformation of a vector from the search space
+        ``x`` to the model parameter space ``p``.
+        """
+        raise NotImplementedError
+
+    def to_search(self, p):
+        """
+        Returns the forward transformation of a vector from the model parameter
+        space ``p`` to the search space ``x``.
+        """
+        raise NotImplementedError
+
+
+class TransformedLogPDF(pints.LogPDF):
+    """
+    A log-PDF that is transformed from the model space to the search space.
```

Comment:
Something like 

``A :class:`pints.LogPDF` that accepts parameters in a transformed search space`` ? (same for others)

Reply:
Done